### ISSUE TITLE
[MJAVADOC-449] javadoc:aggregate fails with JDK 9

### ISSUE THREAD

**Issue by jira-importer** [apache/maven-javadoc-plugin#741]
**[Hervé Boutemy](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hboutemy)** opened **[MJAVADOC-449](https://issues.apache.org/jira/browse/MJAVADOC-449?redirect=false)** and commented

trying `mvn -Preporting site` with Maven core and maven-javadoc-plugin upgraded to 2.10.4-SNAPSHOT to benefit from already JDK 9 related fixed issues shows another one

```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-site-plugin:3.4:site (default-site) on project maven: Error generating maven-javadoc-plugin:2.10.4-SNAPSHOT:aggregate: Unable to resolve artifact:groupId = 'org.apache.maven.plugin-tools'
[ERROR] artifactId = 'maven-plugin-tools-javadoc'
[ERROR] version = '3.4': Missing:
[ERROR] ----------
[ERROR] 1) com.sun:tools:jar:1.4.2
[ERROR] 
[ERROR] Try downloading the file manually from the project website.
[ERROR] 
[ERROR] Then, install it using the command:
[ERROR] mvn install:install-file -DgroupId=com.sun -DartifactId=tools -Dversion=1.4.2 -Dpackaging=jar -Dfile=/path/to/file
[ERROR] 
[ERROR] Alternatively, if you host your own repository you can deploy the file there:
[ERROR] mvn deploy:deploy-file -DgroupId=com.sun -DartifactId=tools -Dversion=1.4.2 -Dpackaging=jar -Dfile=/path/to/file -Durl=[url] -DrepositoryId=[id]
[ERROR] 
[ERROR] Path to dependency:
[ERROR] 1) org.apache.maven.plugin-tools:maven-plugin-tools-javadoc:jar:3.4
[ERROR] 2) com.sun:tools:jar:1.4.2
[ERROR] 
[ERROR] ----------
[ERROR] 1 required artifact is missing.
[ERROR] 
[ERROR] for artifact:
[ERROR] org.apache.maven.plugin-tools:maven-plugin-tools-javadoc:jar:3.4
[ERROR] 
[ERROR] from the specified remote repositories:
[ERROR] apache.snapshots (http://repository.apache.org/snapshots, releases=false, snapshots=true),
[ERROR] central (https://repo.maven.apache.org/maven2, releases=true, snapshots=false)
[ERROR] -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
```

---

**Affects:** 2.10.3

**Attachments:**
- [20180618-MJAVADOC449.zip](https://issues.apache.org/jira/secure/attachment/12928293/20180618-MJAVADOC449.zip) (_7.25 kB_)
- [maven-javadoc-plugin-aggregate-test.zip](https://issues.apache.org/jira/secure/attachment/12910608/maven-javadoc-plugin-aggregate-test.zip) (_3.51 kB_)

**Issue Links:**
- [MPLUGIN-300](https://issues.apache.org/jira/browse/MPLUGIN-300) Prepare taglets for Java9

- [SIS-561](https://issues.apache.org/jira/browse/SIS-561) Upgrade to Java 11
 (_**"is depended upon by"**_)

**Remote Links:**
- [Page
](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=62692896)

3 votes, 14 watchers

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

It's actually not a m-javadoc-p issue, but  a http://maven.apache.org/components/plugin-tools/maven-plugin-tools-javadoc/dependencies.html issue.

**Comment by jira-importer** [comment]
**[Hervé Boutemy](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hboutemy)** commented

if it happens when running javadoc plugin, it's a m-javadoc-p issue: the root cause may be elsewhere, of course, and we may have to release a whole chain of components to get this goal working with Java 9

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

The problem is caused by `org.apache.maven:maven-parent:27`, which adds 2 tagletArtifacts to the configuration o the maven-javadoc-plugin: org.apache.maven.plugin-tools:maven-plugin-tools-javadoc and org.codehaus.plexus:plexus-javadoc. These custom tags depend on tools.jar (as specified in their pom.xml), which is not going to work with Java9. I've ask the jigsaw-dev list how this should work.

**Comment by jira-importer** [comment]
**[Hervé Boutemy](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hboutemy)** commented

thanks for the investigations: it makes sense
and now, with that in mind, it's not a m-javadoc-p issue per se, but good experience on what can go wrong when upgrading :)

**Comment by jira-importer** [comment]
**[Hervé Boutemy](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hboutemy)** commented

while at it, bug found in JDK 9 ea: JDK-8154482:
new doclet should support old taglets
https://bugs.openjdk.java.net/browse/JDK-8154482

see http://mail.openjdk.java.net/pipermail/javadoc-dev/2016-April/000211.html

**Comment by jira-importer** [comment]
**[Gili](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cowwoc)** commented

The aforementioned JDK bug is now marked as fixed.

**Comment by jira-importer** [comment]
**[Karl Heinz Marbaise](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=khmarbaise)** commented

If I correctly understand the comments here: This is fixed ? Can we close this?

**Comment by jira-importer** [comment]
**[James Marble](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=james.d.marble)** commented

I'm getting a different error, but I still cannot run javadoc:aggregate on any multi-module JDK 9 project with at least one module-info.java. It always complains about a missing module descriptor for the root project:

`[ERROR] no module descriptor for maven-javadoc-aggregate-test:maven-javadoc-aggregate-test`

This happens on javadoc:aggregate or a site build that is configured for javadoc:aggregate.

```java
[INFO] Scanning for projects...
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Build Order:
[INFO] 
[INFO] maven-javadoc-aggregate-test
[INFO] maven-javadoc-aggregate-test-child
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] Building maven-javadoc-aggregate-test 0.0.1-SNAPSHOT
[INFO] ------------------------------------------------------------------------
[INFO] 
[INFO] >>> maven-javadoc-plugin:3.0.0:aggregate (default-cli) > compile @ maven-javadoc-aggregate-test >>>
[INFO] 
[INFO] >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
[INFO] Forking maven-javadoc-aggregate-test-child 0.0.1-SNAPSHOT
[INFO] >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
[INFO] 
[INFO] --- maven-resources-plugin:2.6:resources (default-resources) @ maven-javadoc-aggregate-test-child ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] skip non existing resourceDirectory C:\Users\jm6534\Documents\maven-javadoc-plugin-aggregate-test\maven-javadoc-aggregate-test-child\src\main\resources
[INFO] 
[INFO] --- maven-compiler-plugin:3.1:compile (default-compile) @ maven-javadoc-aggregate-test-child ---
[INFO] Changes detected - recompiling the module!
[INFO] Compiling 3 source files to C:\Users\jm6534\Documents\maven-javadoc-plugin-aggregate-test\maven-javadoc-aggregate-test-child\target\classes
[INFO] 
[INFO] <<< maven-javadoc-plugin:3.0.0:aggregate (default-cli) < compile @ maven-javadoc-aggregate-test <<<
[INFO] 
[INFO] --- maven-javadoc-plugin:3.0.0:aggregate (default-cli) @ maven-javadoc-aggregate-test ---
[ERROR] no module descriptor for maven-javadoc-aggregate-test:maven-javadoc-aggregate-test
[INFO] 
Loading source file C:\Users\jm6534\Documents\maven-javadoc-plugin-aggregate-test\maven-javadoc-aggregate-test-child\src\main\java\module-info.java...
1 error
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary:
[INFO] 
[INFO] maven-javadoc-aggregate-test ....................... FAILURE [ 5.050 s]
[INFO] maven-javadoc-aggregate-test-child ................. SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 5.313 s
[INFO] Finished at: 2018-02-14T08:46:47-08:00
[INFO] Final Memory: 20M/65M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-javadoc-plugin:3.0.0:aggregate (default-cli) on project maven-javadoc-aggregate-test: An error has occurred in Javadoc report generation:
[ERROR] Exit code: 1 - error: module not found: maven.javadoc.aggregate.test.child
[ERROR] 
[ERROR] Command line was: "C:\Program Files\Java\jdk-9.0.1\bin\javadoc.exe" @options @packages @argfile
[ERROR] 
[ERROR] Refer to the generated Javadoc files in 'C:\Users\jm6534\Documents\maven-javadoc-plugin-aggregate-test\target\site\apidocs' dir.
[ERROR] -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
```

Attached is a minimal example that produces this error. Deleting module-info.java "fixes" the problem. [^maven-javadoc-plugin-aggregate-test.zip]

**Comment by jira-importer** [comment]
**[Ed Burns](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=edburns@acm.org)** commented

This is still a problem with 3.0.1.  I'll attach my simple module as well, but I see:

`[ERROR] no module descriptor for parent:parent`

The workaround suggested above of removing the **module-info.java** files is not at all satisfying because doing that you lose out on the module metadata presented in Javadoc.

**Comment by jira-importer** [comment]
**[Lewis John McGibbney](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=lewismc)** commented

javadoc:aggregate fails with following JDK

Apache Maven 3.5.4 (1edded0938998edf8bf061f1ceb3cfdeccf443fe; 2018-06-17T11:33:14-07:00)
Maven home: /usr/local/Cellar/maven/3.5.4/libexec
Java version: 10.0.2, vendor: Oracle Corporation, runtime: /Library/Java/JavaVirtualMachines/jdk-10.0.2.jdk/Contents/Home
Default locale: en_US, platform encoding: UTF-8
OS name: "mac os x", version: "10.12.6", arch: "x86_64", family: "Mac"

maven-javadoc-plugin version 3.0.1

**Comment by jira-importer** [comment]
**[Martin Desruisseaux](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=desruisseaux)** commented

I think that the issue is deeper than aggregated javadoc only. In order to generate aggregated javadoc with a front page listing modules, we need to use the `\--module-path` and `\--module-source-path` options of JDK tools (if there is another way, I did not found it yet). But those options require a directory having the exact module name. Current Maven standard directory layout is incompatible with this requirement. This problem impacts no only javadoc, but all JDK tools working with modules (at least `javadoc`, `javac` and `jlink`). For example when used with `javac`, "aggregated compilation" allows to run Annotation Processing Tools on all modules in one pass (useful if the processing tools want to combine information from all modules).

If we need to revisit directory layout in order to support aggregated javadoc, then we could expand the scope for fixing also other issues (in my opinion) in supporting multi-language projects, for example in projects using Java, Python and C/C++ together. A new directory layout [has been proposed on the Maven developer mailing list](https://mail-archives.apache.org/mod_mbox/maven-dev/201808.mbox/%3C91eb6c15-9fc6-6d2a-56c9-962c2be4c227%40geomatys.com%3E). But [implementing and experimenting](https://github.com/Geomatys/maven-plugins/wiki) it would require volunteer time, maybe outside the current Maven maintainers team, until the value of such proposal is demonstrated (or not).

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

Fixed in [1a213f370a0df432bcf82ce440a68c9c4e80eb61](https://gitbox.apache.org/repos/asf?p=maven-javadoc-plugin.git;a=commit;h=1a213f370a0df432bcf82ce440a68c9c4e80eb61)

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

[Nick Williams](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=beamerblvd) Could you try 3.1.0-SNAPSHOT from https://repository.apache.org/content/repositories/snapshots/ ? A lot has been improved the last couple of months. If it is still a problem, please create a new issue.
